activerecord/lib/active_record/relation/batches.rb

```diff
@@ -227,35 +230,36 @@ def in_batches(of: 1000, start: nil, finish: nil, load: false, error_on_ignore:
         batch_limit = remaining if remaining < batch_limit
       end
 
-      relation = relation.reorder(batch_order(order)).limit(batch_limit)
-      relation = apply_limits(relation, start, finish, order)
+
+      relation = relation.reorder(batch_order(order, cursor_column)).limit(batch_limit)
+      relation = apply_limits(relation, start, finish, order, cursor_column)
       relation.skip_query_cache! # Retaining the results in the query cache would undermine the point of batching
       batch_relation = relation
       empty_scope = to_sql == klass.unscoped.all.to_sql
 
       loop do
         if load
           records = batch_relation.records
-          ids = records.map(&:id)
-          yielded_relation = where(primary_key => ids)
+          ids = records.map { |r| r.send(cursor_column) }
+          yielded_relation = where(cursor_column => ids)
           yielded_relation.load_records(records)
         elsif (empty_scope && use_ranges != false) || use_ranges
-          ids = batch_relation.pluck(primary_key)
+          ids = batch_relation.pluck(cursor_column)
           finish = ids.last
           if finish
-            yielded_relation = apply_finish_limit(batch_relation, finish, order)
+            yielded_relation = apply_finish_limit(batch_relation, finish, order, cursor_column)
             yielded_relation = yielded_relation.except(:limit, :order)
             yielded_relation.skip_query_cache!(false)
           end
         else
-          ids = batch_relation.pluck(primary_key)
-          yielded_relation = where(primary_key => ids)
+          ids = batch_relation.pluck(cursor_column)
+          yielded_relation = where(cursor_column => ids)
         end
 
         break if ids.empty?
 
-        primary_key_offset = ids.last
-        raise ArgumentError.new("Primary key not included in the custom select clause") unless primary_key_offset
+        cursor_column_offset = ids.last
+        raise ArgumentError.new("Primary key not included in the custom select clause") unless cursor_column_offset
```

Comment:
We probably want to generalize the error message to say `Cursor column not included in the ...` instead of referencing the PK specifically.